# Cleaning data and setting data types
In our initial exploration we already saw some missing values, duplicated rows, and wrong data types. In this notebook we will clean our data and save those cleaned tables as csv-files.

We will start with orders and orderlines.
## Orders

In [1]:
import pandas as pd

In [2]:
orders = pd.read_csv('data/orders.csv')
orders.head()

order_id         created_date  total_paid      state
0    241319  2017-01-02 13:35:40       44.99  Cancelled
1    241423  2017-11-06 13:10:02      136.15  Completed
2    242832  2017-12-31 17:40:03       15.76  Completed
3    243330  2017-02-16 10:59:38       84.98  Completed
4    243784  2017-11-24 13:35:19      157.86  Cancelled

In [3]:
orders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 226909 entries, 0 to 226908
Data columns (total 4 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   order_id      226909 non-null  int64  
 1   created_date  226909 non-null  object 
 2   total_paid    226904 non-null  float64
 3   state         226909 non-null  object 
dtypes: float64(1), int64(1), object(2)
memory usage: 6.9+ MB


We saw in our initial exploration that the state of the orders with missing *total_paid* values is pending. We therefore do not have to worry about that.

But since we are interested in the dates, we'd prefer that column to be of data type datetime.

In [4]:
orders['created_date'] = pd.to_datetime(orders['created_date'])

In [5]:
min(orders.created_date)

Timestamp('2017-01-01 00:07:19')

In [6]:
max(orders.created_date)

Timestamp('2018-03-14 13:58:36')

In [7]:
orders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 226909 entries, 0 to 226908
Data columns (total 4 columns):
 #   Column        Non-Null Count   Dtype         
---  ------        --------------   -----         
 0   order_id      226909 non-null  int64         
 1   created_date  226909 non-null  datetime64[ns]
 2   total_paid    226904 non-null  float64       
 3   state         226909 non-null  object        
dtypes: datetime64[ns](1), float64(1), int64(1), object(1)
memory usage: 6.9+ MB


We succesfully changed the date to datetime. It is easy to see now that the orders were created between Jan 2017 and Mar 2018.

Next up we'll look at the orderlines table:
## Orderlines

In [8]:
orderlines = pd.read_csv('data/orderlines.csv')

In [9]:
orderlines.head()

id  id_order  product_id  product_quantity      sku unit_price  \
0  1119109    299539           0                 1  OTT0133      18.99   
1  1119110    299540           0                 1  LGE0043     399.00   
2  1119111    299541           0                 1  PAR0071     474.05   
3  1119112    299542           0                 1  WDT0315      68.39   
4  1119113    299543           0                 1  JBL0104      23.74   

                  date  
0  2017-01-01 00:07:19  
1  2017-01-01 00:19:45  
2  2017-01-01 00:20:57  
3  2017-01-01 00:51:40  
4  2017-01-01 01:06:38

In [10]:
orderlines.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 293983 entries, 0 to 293982
Data columns (total 7 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   id                293983 non-null  int64 
 1   id_order          293983 non-null  int64 
 2   product_id        293983 non-null  int64 
 3   product_quantity  293983 non-null  int64 
 4   sku               293983 non-null  object
 5   unit_price        293983 non-null  object
 6   date              293983 non-null  object
dtypes: int64(4), object(3)
memory usage: 15.7+ MB


In [11]:
# orderlines['unit_price'] = pd.to_numeric(orderlines['unit_price'])

When trying to convert the _unit_price_ to a numeric value, we get an error message. This pops up because some of these values have multiple points in it. This was probably done for readability of big numbers, but for our purposes we have to get rid of it. What we want to do is to remove every dot except for the last one using regular expressions

In [12]:
orderlines.loc[orderlines.unit_price.str.count('\.') >2]

Empty DataFrame
Columns: [id, id_order, product_id, product_quantity, sku, unit_price, date]
Index: []

We see that these numbers have at most 2 dots. So if we remove the first dot of those fields, we'll be fine.

In [13]:
orderlines.loc[orderlines.unit_price.str.count('\.') >1, 'unit_price'] = orderlines.loc[orderlines.unit_price.str.count('\.') >1, 'unit_price'].str.replace('\.', '', n=1, regex=True)

In [14]:
orderlines.head(10)

id  id_order  product_id  product_quantity      sku unit_price  \
0  1119109    299539           0                 1  OTT0133      18.99   
1  1119110    299540           0                 1  LGE0043     399.00   
2  1119111    299541           0                 1  PAR0071     474.05   
3  1119112    299542           0                 1  WDT0315      68.39   
4  1119113    299543           0                 1  JBL0104      23.74   
5  1119114    295310           0                10  WDT0249     231.79   
6  1119115    299544           0                 1  APP1582    1137.99   
7  1119116    299545           0                 1  OWC0100      47.49   
8  1119119    299546           0                 1  IOT0014      18.99   
9  1119120    295347           0                 1  APP0700      72.19   

                  date  
0  2017-01-01 00:07:19  
1  2017-01-01 00:19:45  
2  2017-01-01 00:20:57  
3  2017-01-01 00:51:40  
4  2017-01-01 01:06:38  
5  2017-01-01 01:14:27  
6  2017-01-01 01:17:21  
7  2017-01-01 01:46:16  
8  2017-01-01 01:50:34  
9  2017-01-01 01:54:11

In [15]:
orderlines['unit_price'] = pd.to_numeric(orderlines['unit_price'])

Looks like it worked! Note that the string method `count` uses regular expressions by default, but for the `replace` method we have to say that we are dealing with regex.

Now to the datetime:

In [16]:
orderlines['date'] = pd.to_datetime(orderlines['date'])

In [17]:
orderlines.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 293983 entries, 0 to 293982
Data columns (total 7 columns):
 #   Column            Non-Null Count   Dtype         
---  ------            --------------   -----         
 0   id                293983 non-null  int64         
 1   id_order          293983 non-null  int64         
 2   product_id        293983 non-null  int64         
 3   product_quantity  293983 non-null  int64         
 4   sku               293983 non-null  object        
 5   unit_price        293983 non-null  float64       
 6   date              293983 non-null  datetime64[ns]
dtypes: datetime64[ns](1), float64(1), int64(4), object(1)
memory usage: 15.7+ MB


## Products

Next we'll tackle the table with the most issues, which is the `products` table. From the initial exploration we know that we have to deal with missing prices, missing descriptions, duplicate rows, and some problems with the actual prices (most of them are way too high), which might be due to an error of wrong data imputation.



In [18]:
products = pd.read_csv('data/products.csv')

In [19]:
products.head()

sku                                           name  \
0  RAI0007              Silver Rain Design mStand Support   
1  APP0023              Apple Mac Keyboard Keypad Spanish   
2  APP0025               Mighty Mouse Apple Mouse for Mac   
3  APP0072  Apple Dock to USB Cable iPhone and iPod white   
4  KIN0007    Mac Memory Kingston 2GB 667MHz DDR2 SO-DIMM   

                                                desc  price promo_price  \
0       Aluminum support compatible with all MacBook  59.99     499.899   
1          USB ultrathin keyboard Apple Mac Spanish.     59     589.996   
2                             mouse Apple USB cable.     59     569.898   
3              IPhone dock and USB Cable Apple iPod.     25     229.997   
4  2GB RAM Mac mini and iMac (2006/07) MacBook Pr...  34.99       31.99   

   in_stock      type  
0         1      8696  
1         0  13855401  
2         0      1387  
3         0      1230  
4         1      1364

#### Duplicates

In [20]:
products.duplicated().sum()

8746

In [21]:
products.drop_duplicates(inplace=True)

In [22]:
products.duplicated(subset = ['sku']).sum()

1

Looking at the duplicates: There are 8746 entries that match some other entry in every column.
If we look at duplicates in the sku-column only, we see that there is one more duplicated entry. Let's look at that one and see which one we want to keep.

In [23]:
products.loc[products.duplicated(subset = ['sku']) == True]

sku                                               name  \
8000  APP1197  Apple iMac 21.5 "Core i5 31 GHz Retina display...   

                                                   desc price promo_price  \
8000  Desktop Apple iMac 21.5 inch i5 31 GHz Retina ...   NaN     1305.59   

      in_stock  type  
8000         0  1282

In [24]:
products.loc[products.sku == 'APP1197']

sku                                               name  \
7992  APP1197  Apple iMac 21.5 "Core i5 31 GHz Retina display...   
8000  APP1197  Apple iMac 21.5 "Core i5 31 GHz Retina display...   

                                                   desc price promo_price  \
7992  Desktop Apple iMac 21.5 inch i5 31 GHz Retina ...  1729     1305.59   
8000  Desktop Apple iMac 21.5 inch i5 31 GHz Retina ...   NaN     1305.59   

      in_stock  type  
7992         0  1282  
8000         0  1282

We found the last duplicate sku APP1197. the second does not have a price, everything else is the same. So we can drop it.

In [25]:
products.drop_duplicates(subset=['sku'], inplace=True)
#products.loc[products.sku == 'APP1197']

#### Data types and missing values

In [26]:
products.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10579 entries, 0 to 19325
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   sku          10579 non-null  object
 1   name         10579 non-null  object
 2   desc         10572 non-null  object
 3   price        10534 non-null  object
 4   promo_price  10579 non-null  object
 5   in_stock     10579 non-null  int64 
 6   type         10529 non-null  object
dtypes: int64(1), object(6)
memory usage: 661.2+ KB


In [27]:
# products['promo_price'] = pd.to_numeric(products['promo_price'])

As it was the case with the orderlines dataframe, there are some issues with multiple dots for some `price`'s and `promo_price`'s.
Fortunately, we already know how to deal with that.

In [28]:
#products.loc[products.promo_price.str.count('\.') >2]
products.loc[products.price.str.count('\.') >2]

Empty DataFrame
Columns: [sku, name, desc, price, promo_price, in_stock, type]
Index: []

In [29]:
products.loc[products.price.str.count('\.') >1, 'price'] = (
    products.loc[products.price.str.count('\.') >1, 
                   'price'].str.replace('\.', '', n=1, regex=True)
)

products.loc[products.promo_price.str.count('\.') >1, 'promo_price'] = (
    products.loc[products.promo_price.str.count('\.') >1, 
                   'promo_price'].str.replace('\.', '', n=1, regex=True)
)

In [30]:
products['price'] = pd.to_numeric(products['price'])
products['promo_price'] = pd.to_numeric(products['promo_price'])

In [31]:
products.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10579 entries, 0 to 19325
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   sku          10579 non-null  object 
 1   name         10579 non-null  object 
 2   desc         10572 non-null  object 
 3   price        10534 non-null  float64
 4   promo_price  10579 non-null  float64
 5   in_stock     10579 non-null  int64  
 6   type         10529 non-null  object 
dtypes: float64(2), int64(1), object(4)
memory usage: 661.2+ KB


For the purpose of later categorization, we'll copy-paste the name of the product into the description for those products that are missing descriptions.
This avoids having to deal with nan-values.

In [32]:
products.loc[products.desc.isna(), 'desc'] = products.loc[products.desc.isna(), 'name']

Looking at the type, we see that all these contain some positive numbers. Since we will not look into the details of it right now (these are likely some company internal numbers), and since not too many products are missing type numbers, we'll simply impute 0.
As it does not make sense to do calculations with these numbers, it is okay for now to keep it as data type `object`.

We might come back to it for categorization later, so we won't drop the column.

In [33]:
products.type.unique()

array(['8696', '13855401', '1387', '1230', '1364', '1325', '5384', '1334',
       '13005399', '13835403', '12995397', '11865403', '13955395', '1216',
       '12355400', '5,49E+11', '1276', '11905404', '12635403', '12755395',
       '1296', '12285400', '1229', '11935397', '12655397', '1404',
       '101781405', '4259', '14035403', '12085400', '1282', '12175397',
       '1424', '9094', '1405', '57445397', nan, '14305406', '10142',
       '12645406', '10230', '12215397', '11821715', '13555403',
       '1,44E+11', '14365395', '5405', '5395', '5398', '21485407',
       '20642062', '1,46E+11', '1280', '12575403', '1433', '1515', '5720',
       '1298', '13615399', '12585395', '1392', '1231', '15435404', '1375',
       '42945397', '12141714', '42931714', '54025401', '21535407', '1416',
       '2,17E+11', '24215399', '11434', '2425', '2434', '2449',
       '13621714', '24861714', '24821716', '5,74E+15', '54864259',
       '51861714', '1714', '51601716', '51871714', '5403', '54085407',
       '2

In [34]:
products.loc[products.type.isna(), 'type'] = 0

In [35]:
products.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10579 entries, 0 to 19325
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   sku          10579 non-null  object 
 1   name         10579 non-null  object 
 2   desc         10579 non-null  object 
 3   price        10534 non-null  float64
 4   promo_price  10579 non-null  float64
 5   in_stock     10579 non-null  int64  
 6   type         10579 non-null  object 
dtypes: float64(2), int64(1), object(4)
memory usage: 661.2+ KB


#### The price issue
Now we want to get rid of our biggest problem: the weird prices.
As mentioned before there is something wrong with a lot of prices and promo_prices given in the table. 
Unfortunately we have no access to the person managing that dataframe and no internal price catalog.
So to fix the prices we assumed two things: 

- The problem with the pricing is caused by some floating point error, therefore the difference of the actual price and the price given in the table can be fixed with multiplication by some order of 10
- The actual price are comparable to product prices that can be found in other online stores

So through extensive online research and using the difference between `price` and `promo_price` as an auxiliary column the prices got fixed manually.

One thing to note: Only the `price` column gets fixed! Mainly because of the following reasons:
- When calculating the total price of an order later, we'll use only the price from the price-column. The promo_price is only used as a reference to find the actual price of the products.
- The promo_price given reflects a discount given by the respective company, probably for special promotional events. The actual vendors might give more or less discount, depending on various circumstances such as: Special dates (Black Friday, Christmas), getting rid of remaining stock, etc.
- One can basically see the discount when comparing the payment of the orders and the total price of the orders based on the prices given in the products dataframe. It's is not 100% accurate because of shipping costs etc., but it's good enough for our puposes.

Another thing to note: It is possible to lose some accuracy of the prices and spend less time on the cleaning process is exchange. This can be done via the use of functions and automizing the process. It is not done here though, since the amount of work was manageable in the timeframe given.

In [36]:
# code here
products.loc[products.price.isna() == True]

sku                                               name  \
34       TWS0019  Twelve South MagicWand support Apple Magic Tra...   
1900     AII0008           Aiino Case MacBook Air 11 '' Transparent   
2039     CEL0020  Celly Ambo Luxury Leather Case + iPhone 6 Case...   
2042     CEL0007  Celly Wallet Case with removable cover Black i...   
2043     CEL0012            Celly Silicone Hard Shell iPhone 6 Blue   
2044     CEL0014        Celly Silicone Hard Shell iPhone 6 Amarillo   
2049     CEL0015    Celly fur-lined Powerbank battery 4000mAh Black   
2051     CEL0018     Celly Wallet Leather Case cover Black iPhone 6   
2052     CEL0023  Celly Ambo Luxury Leather Case + Case Gold iPh...   
2053     CEL0025  Celly Ambo Luxury Leather Case + Case iPhone 6...   
12052    CEL0028  Celly Swing Arm button selfie with Bluetooth B...   
12053    CEL0029  Celly extensible arm selfie with Bluetooth but...   
12054    CEL0030  Celly Smart Drive Housing + clip car holder iP...   
12055    CEL0031  Celly Smart Drive Case + Car clip support iPho...   
12059    CEL0032      Celly Gost clamp fixing car + Magnetica Black   
12060    CEL0042  Celly iPhone Case Fluor 6s / 6 Black / Transpa...   
12061    CEL0043  Celly Fluor Case iPhone 6 / 6S Pink / Transparent   
12062    CEL0044  Celly iPhone Case Fluor 6s / 6 Yellow / Transp...   
12064    CEL0038           Celly Frost Cover for iPhone 6 / 6S Rosa   
12067    CEL0033  Celly Waterproof iPad Case / Air / Pro 97 "Tra...   
12068    CEL0034  Celly Ambo cover with lid iPhone 6 / 6S Plus B...   
12069    CEL0035  Celly Ambo cover with lid iPhone 6 / 6S Plus B...   
12070    CEL0036  Celly Adapter Micro USB connector to Lightning...   
12071    CEL0051   Celly 4000mAh Battery Power Bank Aluminum Dorado   
12072    CEL0052   Celly 4000mAh Battery Power Bank Aluminum Silver   
12073    CEL0053             Celly IceCube Case iPhone 6 / 6S Black   
12074    CEL0054              Celly IceCube Case iPhone 6 / 6S Rosa   
12075    CEL0055             Celly IceCube Case iPhone 6 / 6S White   
12076    CEL0056          IceCube Celly iPhone Case 6 / 6s Amarillo   
12077    CEL0057        Celly IceCube Case iPhone 6 / 6S Plus Black   
12078    CEL0058         Celly IceCube Case iPhone 6 / 6S Plus Rosa   
12079    CEL0059        Celly IceCube Case iPhone 6 / 6S Plus White   
12080    CEL0060     Celly IceCube Case iPhone 6 / 6S Plus Amarillo   
12088    CEL0040     Celly Frost Cover for iPhone 6 / 6S Plus White   
12089    CEL0041              Celly iPhone 6S Plus Frost Case Black   
12092    CEL0045         Laser Case Celly iPhone 6 / 6S Dark Silver   
12093    CEL0046        Laser Case Celly iPhone 6 / 6S Clear Silver   
12095    CEL0047              Laser Case Celly iPhone 6 / 6S Dorado   
12096    CEL0048                Laser Case Celly iPhone 6 / 6S Rosa   
12097    CEL0049         Laser Case Celly iPhone 6 / 6S Plus Dorado   
12098    CEL0050   Laser Case Celly iPhone 6 / 6S Plus Silver Clear   
12100    CEL0037          Celly Frost Cover for iPhone 6 / 6S Black   
12238    CEL0027          Celly Frost Cover for iPhone 6 / 6S White   
12242    CEL0039       Celly Frost Cover for iPhone 6 / 6S Amarillo   
12947  CEL0034-A    (Open) Celly iPhone 6S Plus Housing Frost White   

                                                    desc  price  promo_price  \
34     MagicWand for wireless keyboard and Magic Trac...    NaN      299.899   
1900       MacBook Air 11-inch casing with matte finish.    NaN       22.990   
2039   Cover and housing together with magnet for iPh...    NaN      399.905   
2042              Case Book for iPhone 6 card case type.    NaN      128.998   
2043                       Hard Shell Silicone iPhone 6.    NaN        4.990   
2044                       Hard Shell Silicone iPhone 6.    NaN       59.895   
2049   Leather-wrapped External Battery 4000mAh for i...    NaN      239.895   
2051   Card case with transparent protective cover fo...    NaN      294.877   
2052   Cover and housing t

we have several entries with missing price. there is a promo price, but it seems that it is 1 order higher in most cases. I would say CEL0012, AII0008 and TWS0019 are the only correct ones there. 

In [37]:
products['no_price'] = products.price.isna()
products.loc[products.no_price]

sku                                               name  \
34       TWS0019  Twelve South MagicWand support Apple Magic Tra...   
1900     AII0008           Aiino Case MacBook Air 11 '' Transparent   
2039     CEL0020  Celly Ambo Luxury Leather Case + iPhone 6 Case...   
2042     CEL0007  Celly Wallet Case with removable cover Black i...   
2043     CEL0012            Celly Silicone Hard Shell iPhone 6 Blue   
2044     CEL0014        Celly Silicone Hard Shell iPhone 6 Amarillo   
2049     CEL0015    Celly fur-lined Powerbank battery 4000mAh Black   
2051     CEL0018     Celly Wallet Leather Case cover Black iPhone 6   
2052     CEL0023  Celly Ambo Luxury Leather Case + Case Gold iPh...   
2053     CEL0025  Celly Ambo Luxury Leather Case + Case iPhone 6...   
12052    CEL0028  Celly Swing Arm button selfie with Bluetooth B...   
12053    CEL0029  Celly extensible arm selfie with Bluetooth but...   
12054    CEL0030  Celly Smart Drive Housing + clip car holder iP...   
12055    CEL0031  Celly Smart Drive Case + Car clip support iPho...   
12059    CEL0032      Celly Gost clamp fixing car + Magnetica Black   
12060    CEL0042  Celly iPhone Case Fluor 6s / 6 Black / Transpa...   
12061    CEL0043  Celly Fluor Case iPhone 6 / 6S Pink / Transparent   
12062    CEL0044  Celly iPhone Case Fluor 6s / 6 Yellow / Transp...   
12064    CEL0038           Celly Frost Cover for iPhone 6 / 6S Rosa   
12067    CEL0033  Celly Waterproof iPad Case / Air / Pro 97 "Tra...   
12068    CEL0034  Celly Ambo cover with lid iPhone 6 / 6S Plus B...   
12069    CEL0035  Celly Ambo cover with lid iPhone 6 / 6S Plus B...   
12070    CEL0036  Celly Adapter Micro USB connector to Lightning...   
12071    CEL0051   Celly 4000mAh Battery Power Bank Aluminum Dorado   
12072    CEL0052   Celly 4000mAh Battery Power Bank Aluminum Silver   
12073    CEL0053             Celly IceCube Case iPhone 6 / 6S Black   
12074    CEL0054              Celly IceCube Case iPhone 6 / 6S Rosa   
12075    CEL0055             Celly IceCube Case iPhone 6 / 6S White   
12076    CEL0056          IceCube Celly iPhone Case 6 / 6s Amarillo   
12077    CEL0057        Celly IceCube Case iPhone 6 / 6S Plus Black   
12078    CEL0058         Celly IceCube Case iPhone 6 / 6S Plus Rosa   
12079    CEL0059        Celly IceCube Case iPhone 6 / 6S Plus White   
12080    CEL0060     Celly IceCube Case iPhone 6 / 6S Plus Amarillo   
12088    CEL0040     Celly Frost Cover for iPhone 6 / 6S Plus White   
12089    CEL0041              Celly iPhone 6S Plus Frost Case Black   
12092    CEL0045         Laser Case Celly iPhone 6 / 6S Dark Silver   
12093    CEL0046        Laser Case Celly iPhone 6 / 6S Clear Silver   
12095    CEL0047              Laser Case Celly iPhone 6 / 6S Dorado   
12096    CEL0048                Laser Case Celly iPhone 6 / 6S Rosa   
12097    CEL0049         Laser Case Celly iPhone 6 / 6S Plus Dorado   
12098    CEL0050   Laser Case Celly iPhone 6 / 6S Plus Silver Clear   
12100    CEL0037          Celly Frost Cover for iPhone 6 / 6S Black   
12238    CEL0027          Celly Frost Cover for iPhone 6 / 6S White   
12242    CEL0039       Celly Frost Cover for iPhone 6 / 6S Amarillo   
12947  CEL0034-A    (Open) Celly iPhone 6S Plus Housing Frost White   

                                                    desc  price  promo_price  \
34     MagicWand for wireless keyboard and Magic Trac...    NaN      299.899   
1900       MacBook Air 11-inch casing with matte finish.    NaN       22.990   
2039   Cover and housing together with magnet for iPh...    NaN      399.905   
2042              Case Book for iPhone 6 card case type.    NaN      128.998   
2043                       Hard Shell Silicone iPhone 6.    NaN        4.990   
2044                       Hard Shell Silicone iPhone 6.    NaN       59.895   
2049   Leather-wrapped External Battery 4000mAh for i...    NaN      239.895   
2051   Card case with transparent protective cover fo...    NaN      294.877   
2052   Cover and housing t

In [38]:
products.loc[products.no_price, 'price'] = products.loc[products.no_price].promo_price

In [39]:
products['price'] = pd.to_numeric(products.price)
products.loc[products.no_price]

sku                                               name  \
34       TWS0019  Twelve South MagicWand support Apple Magic Tra...   
1900     AII0008           Aiino Case MacBook Air 11 '' Transparent   
2039     CEL0020  Celly Ambo Luxury Leather Case + iPhone 6 Case...   
2042     CEL0007  Celly Wallet Case with removable cover Black i...   
2043     CEL0012            Celly Silicone Hard Shell iPhone 6 Blue   
2044     CEL0014        Celly Silicone Hard Shell iPhone 6 Amarillo   
2049     CEL0015    Celly fur-lined Powerbank battery 4000mAh Black   
2051     CEL0018     Celly Wallet Leather Case cover Black iPhone 6   
2052     CEL0023  Celly Ambo Luxury Leather Case + Case Gold iPh...   
2053     CEL0025  Celly Ambo Luxury Leather Case + Case iPhone 6...   
12052    CEL0028  Celly Swing Arm button selfie with Bluetooth B...   
12053    CEL0029  Celly extensible arm selfie with Bluetooth but...   
12054    CEL0030  Celly Smart Drive Housing + clip car holder iP...   
12055    CEL0031  Celly Smart Drive Case + Car clip support iPho...   
12059    CEL0032      Celly Gost clamp fixing car + Magnetica Black   
12060    CEL0042  Celly iPhone Case Fluor 6s / 6 Black / Transpa...   
12061    CEL0043  Celly Fluor Case iPhone 6 / 6S Pink / Transparent   
12062    CEL0044  Celly iPhone Case Fluor 6s / 6 Yellow / Transp...   
12064    CEL0038           Celly Frost Cover for iPhone 6 / 6S Rosa   
12067    CEL0033  Celly Waterproof iPad Case / Air / Pro 97 "Tra...   
12068    CEL0034  Celly Ambo cover with lid iPhone 6 / 6S Plus B...   
12069    CEL0035  Celly Ambo cover with lid iPhone 6 / 6S Plus B...   
12070    CEL0036  Celly Adapter Micro USB connector to Lightning...   
12071    CEL0051   Celly 4000mAh Battery Power Bank Aluminum Dorado   
12072    CEL0052   Celly 4000mAh Battery Power Bank Aluminum Silver   
12073    CEL0053             Celly IceCube Case iPhone 6 / 6S Black   
12074    CEL0054              Celly IceCube Case iPhone 6 / 6S Rosa   
12075    CEL0055             Celly IceCube Case iPhone 6 / 6S White   
12076    CEL0056          IceCube Celly iPhone Case 6 / 6s Amarillo   
12077    CEL0057        Celly IceCube Case iPhone 6 / 6S Plus Black   
12078    CEL0058         Celly IceCube Case iPhone 6 / 6S Plus Rosa   
12079    CEL0059        Celly IceCube Case iPhone 6 / 6S Plus White   
12080    CEL0060     Celly IceCube Case iPhone 6 / 6S Plus Amarillo   
12088    CEL0040     Celly Frost Cover for iPhone 6 / 6S Plus White   
12089    CEL0041              Celly iPhone 6S Plus Frost Case Black   
12092    CEL0045         Laser Case Celly iPhone 6 / 6S Dark Silver   
12093    CEL0046        Laser Case Celly iPhone 6 / 6S Clear Silver   
12095    CEL0047              Laser Case Celly iPhone 6 / 6S Dorado   
12096    CEL0048                Laser Case Celly iPhone 6 / 6S Rosa   
12097    CEL0049         Laser Case Celly iPhone 6 / 6S Plus Dorado   
12098    CEL0050   Laser Case Celly iPhone 6 / 6S Plus Silver Clear   
12100    CEL0037          Celly Frost Cover for iPhone 6 / 6S Black   
12238    CEL0027          Celly Frost Cover for iPhone 6 / 6S White   
12242    CEL0039       Celly Frost Cover for iPhone 6 / 6S Amarillo   
12947  CEL0034-A    (Open) Celly iPhone 6S Plus Housing Frost White   

                                                    desc    price  \
34     MagicWand for wireless keyboard and Magic Trac...  299.899   
1900       MacBook Air 11-inch casing with matte finish.   22.990   
2039   Cover and housing together with magnet for iPh...  399.905   
2042              Case Book for iPhone 6 card case type.  128.998   
2043                       Hard Shell Silicone iPhone 6.    4.990   
2044                       Hard Shell Silicone iPhone 6.   59.895   
2049   Leather-wrapped External Battery 4000mAh for i...  239.895   
2051   Card case with transparent protective cover fo...  294.877   
2052   Cover and housing together with magnet for iPh...  329.894   
2053   Cover and housing together with magnet for iPh...

In [40]:
products.loc[(products.promo_price > products.price), 'promo_price'] /= 10

In [41]:
#products['promo_price'] = round(products.promo_price,2)
#products['price'] = round(products.price,2)
#products.loc[products.promo_price > products.price]

In [42]:
products.loc[(products.no_price), 'price'] /= 10
products.loc[(products.no_price), 'promo_price'] /= 10

In [43]:
products.loc[products.sku.isin(['TWS0019', 'AII0008', 'CEL0012']), 'price'] *= 10
products.loc[products.sku.isin(['TWS0019', 'AII0008', 'CEL0012']), 'promo_price'] *= 10

In [44]:
products['price_promo_difference'] = products.price - products.promo_price
products.loc[products.price < products.promo_price].sort_values('price_promo_difference').head(50)

sku                                               name  \
14718  QNA0189        QNAP TVS-1271U-RP NAS with redundant source   
14717  QNA0188  QNAP TVS-1271U-RP NAS server with redundant po...   
14716  QNA0187  QNAP TVS-1271U-RP NAS server with redundant po...   
15111  EIZ0022  CS2730 Monitor Eizo ColorEdge 27 "QHD AdobeRGB...   
11060  PAC1388     Pack QNAP TS-251 + | 2GB RAM | WD 16TB Network   
14719  QNA0190  QNAP TVS-871U-RP NAS Server | Core i5 | 8GB RA...   
1749   PAC0607         Synology DS215j Pack | 8TB Seagate Desktop   
2879   APP1084                       Apple iPod Touch 64GB Silver   
2880   APP1087                         Apple iPod Touch 64GB Gold   
2881   APP1090                   Apple iPod Touch 64GB Space Gray   
2882   APP1093                         Apple iPod Touch 64GB Rosa   
2883   APP1096                         Apple iPod Touch 64GB Blue   
11288  OWC0132            OWC Aura 2TB SSD for Apple Mac Pro 2013   
968    DRO0022               Connected Cloud NAS Data Transporter   
12279  LAC0184                 Fuel Drive USB 3.0 LaCie 1TB Wi-Fi   
2498   ADO0075              Adobe Premiere Pro CC -Edition video-   
2505   ADO0085                 Adobe Photoshop CC -Edition image-   
376    REP0100               Color change to White Apple iPhone 4   
377    REP0052               Color change to White Apple iPhone 4   
1287   MOS0153  Moshi Venturo shoulder bag / backpack Macbook ...   
1807   SAN0069                   SanDisk Ultra II SSD 240GB Drive   
12853  SAN0136          Plus 240 GB SanDisk SSD 25 "SATA 6 GB / s   
12852  SAN0137            Plus 120GB SanDisk SSD 25 "SATA 6Gb / s   
11404  DLK0128               D-Link DES-1005P PoE Ethernet Switch   
674    GRT0295                    Griffin iPad Kiosk Stand Safety   
13148  SYN0158             Synology MailPlus License for 20 users   
11674  KIN0156    Mac memory Kingston 4GB SO-DIMM DDR3åÊ1333MhzåÊ   
11453  DLK0125           D-Link GO-SW-24G Gigabit Ethernet Switch   
929    LAC0103                      LaCie CloudBox 2TB Hard Drive   
17405  OWC0248  OWC SSD expansion Aura 2TB 6G MacBook Pro 2012...   
181    APP0574  Apple MacBook Pro 133 '' i7 29GHz | 4GB RAM | ...   
2971   SEA0091  Seagate Expansion 1TB Hard Drive 2.5 "USB 3.0 ...   
11446  DLK0123        D-Link GO-RT-N300 Wireless Router N300 Easy   
11395  DLK0110  D-Link DIR-880L Router AC1900 Wi-Fi Dual-Band ...   
561    KIN0079                       Kingston V300 SSD Disk 240GB   
10993  BNQ0044                 LED Monitor Benq GW2270H 215 "HDMI   
468    OWC0056     Mac OWC Memory 4GB (2x2GB) 667MHz DDR2 FB-DIMM   
2890   APP1102                      Apple iPod Shuffle 2GB Silver   
2699   APP1097                        Apple iPod Nano 16Gb Silver   
2885   APP1099                    Apple iPod Nano 16Gb Space Gray   
2870   APP1098                          Apple iPod Nano 16Gb Gold   
2887   APP1101                          Apple iPod Nano 16Gb Blue   
2886   APP1100                          Apple iPod Nano 16Gb Rosa   
11813  APP1488            Apple iPad Pro 9.7 "32GB Wi-Fi Oro Rosa   
13324  APP1756       Apple iPad Mini 4 Wi-Fi + Cellular 32GB Gold   
13321  APP1754  Apple iPad Mini 4 Wi-Fi + Cellular 32GB Space ...   
11821  APP1497  Apple iPad Pro 9.7 "Wi-Fi + Cellular 32GB Spac...   
11823  APP1499     Apple iPad Pro 9.7 "Wi-Fi + Cellular 32GB Gold   
11824  APP1500  Apple iPad Pro 9.7 "Wi-Fi + Cellular 32GB Rose...   
11812  APP1487                 Apple iPad Pro 9.7 "32GB Wi-Fi Oro   

                                                    desc     price  \
14718  Expansion unit rack bays 12 format and 8GB of ...  2346.190   
14717  Expansion unit rack bays 12 and format memory ...  2781.790   
14716  Expansion unit rack format with 12 bays and 32 GB  3386.790   
15111  Professional monitor for color management Adob...  1034.990   
11060  Nas QNAP TS-251 + 2GB RAM memory + 16TB (2x8TB...   673.702   
14719          NAS system format rack 8 bays and 8GB RAM  2418.790   
1

Looking at the price_promo_difference, we see that this is enormous. Prices and (mostly) promo_prices are way too high. Sometimes we see a reasonable promo_price for a macbook, so many of these entries have to be divided by 100. In a first step I go with a price diff of 40000

In [45]:
products.sort_values('price_promo_difference', ascending=False).head(50)

sku                                               name  \
16608    APP2333  Apple MacBook Pro 15 "Core i7 Touch Bar 31GHz ...   
16824    APP2299  Apple MacBook Pro 13 "Core i7 Touch Bar 35GHz ...   
16836    APP2310  Apple MacBook Pro 13 "Core i7 Touch Bar 35GHz ...   
16834    APP2307  Apple MacBook Pro 13 "Core i5 Touch Bar 33GHz ...   
16812    APP2296  Apple MacBook Pro 13 "Core i5 Touch Bar 33GHz ...   
16838    APP2312  Apple MacBook Pro 13 "Core i5 Touch Bar 31GHz ...   
16828    APP2301  Apple MacBook Pro 13 "Core i5 Touch Bar 31GHz ...   
16832    APP2303  Apple MacBook Pro 13 "Core i5 Touch Bar 33GHz ...   
16840    APP2314  Apple MacBook Pro 13 "Core i5 Touch Bar 33GHz ...   
17487    APP2490                         Apple iPhone 64GB X Silver   
16127  WIT0011-A  Open - Withings Blood Pressure Monitor Wireles...   
19248    DJI0026     DJI Mavic Air Drone cuadricóptero Arctic White   
19249    DJI0025       DJI Mavic Air Drone Black Onyx cuadricóptero   
11822    APP1498   Apple iPad Pro 9.7 "Wi-Fi + Cellular 32GB Silver   
898      REP0156                        iPhone 5 GSM antenna repair   
941      REP0185                       Home button repair iPad mini   
11765    APP1477                       Apple iPhone 16GB Space Gray   
12301    APP1603       Space Gray Apple Watch Sport 42mm Black Belt   
19252    LIN0015          Linksys Wi-Fi Velop System AC6600 3 units   
15992    PAC1983  Synology DS1817 + | 16GB RAM | 64TB (8x8TB) Se...   
19181    UBI0007  Ubiquiti Amplifi Wi-Fi Mesh Router + 2 Mesh Ac...   
15922    PAC2313  Synology DS1517 + NAS Server | 8GB RAM | 60TB ...   
1809     SAN0071                   SanDisk Ultra II SSD 960GB Drive   
15919    PAC2023  Synology DS1517 + | 16GB RAM | 30TB (5x6TB) WD...   
19250    LIN0013           Linksys Wi-Fi Velop system AC2200 1 unit   
15401    REP0350                        Full Screen Repair iPad Air   
19173    RIN0007              Camera Intercom Video Doorbell Ring 2   
17486    APP2494                                   Apple TV 4K 64GB   
14784    PAC1715  Pack QNAP TS-451A NAS 8GB RAM l l 24TB (4x6TB)...   
17030    TAD0008  Tado thermostat Intelligent Bridge + Starter K...   
15980    PAC1971  Synology DS1817 + | 8GB RAM | 64TB (8x8TB) Sea...   
11258    PAC1675          QNAP TS-453A | 16GB | 12TB (4x3TB) WD Red   
16001    PAC2309  Synology DS1817 + | 16GB RAM | 96TB (8x12TB) S...   
16000    PAC2307  Synology DS1817 + | 8GB RAM | 96TB (8x12TB) Se...   
15981    PAC1972  Synology DS1817 + | 8GB RAM | 80TB (8x10TB) Se...   
15993    PAC1984  Synology DS1817 + | 16GB RAM | 80TB (8x10TB) S...   
15429    REP0339                         Repair Touch Screen iPad 4   
19238    WAC0248                 Wacom support for MobileStudio Pro   
12884    PAC2041  Synology DS216 + II | Seagate 4TB IronWolf for...   
15998    PAC1989  Synology DS1817 + | 16GB RAM | 64TB (8x8TB) WD...   
15986    PAC1977  Synology DS1817 + | 8GB RAM | 64TB (8x8TB) WD Red   
14786    PAC1716  Pack QNAP TS-451A NAS 8GB RAM l l 40TB (4x10TB...   
2906     SNN0051                  Sonnet Tempo card PCIe SSD 512 Gb   
12872    PAC2029  Synology DS216 + II | Seagate 4TB IronWolf for...   
16003    PAC2308  Synology DS1817 + | 8GB RAM | 80TB (8x10TB) WD...   
16004    PAC2310  Synology DS1817 + | 16GB RAM | 80TB (8x10TB) W...   
17673  APP1727-A  Open - Series 2 42mm Apple Watch Stainless Ste...   
14566    PAC1559  Enlargement Kit 2TB SSD Crucial MX300 + RAM + ...   
15979    PAC1970  Synology DS1817 + | 8GB RAM | 48TB (8x6TB) Sea...   
15991    PAC1982  Synology DS1817 + | 16GB RAM | 48TB (8x6TB) Se...   

                                                    desc       price  \
16608  New MacBook Pro 15-inch Core i7 Touch Bar 31Gh...  352559.394   
16824  New MacBook Pro 13 inch Touch Bar 35 GHz Core ...  336559.443   
16836  New MacBook Pro 13 inch Touch Bar 35 GHz Core ...  336559.443   
16834  New MacBook Pro 13-inch Core i5 Touch Bar 33 G...  336559.443   
16812  New MacBook Pro 13-inch 

In [46]:
products.loc[products.price_promo_difference > 40000, 'price'] /= 100

Now we take a closer look to the macbooks

In [47]:
products.loc[products.name.str.contains('MacBook')].head(5)

sku                                               name  \
13  APP0156              45W MagSafe charger Apple MacBook Air   
15  MOS0021               Clearguard Moshi MacBook Pro and Air   
16  APP0206  60W MagSafe charger Apple MacBook and MacBook ...   
17  HGD0007       Henge Docks Docking Station MacBook Pro 13 "   
20  APP0247    85W MagSafe charger Apple MacBook Pro 15 "/ 17"   

                                                 desc  price  promo_price  \
13     Charger MacBook Air MagSafe 45 W (MC747Z / A).  89.00      84.9904   
15  Keyboard Protector MacBook Pro 13-inch Retina ...  24.95      24.9889   
16  Charger MacBook and 13 inch MacBook MagSafe 60...  89.00      84.9904   
17  Dock and support with multiple connections for...  69.99      69.9864   
20     Charger MacBook Pro MagSafe 85 W (MC556Z / A).  89.00      84.9904   

    in_stock      type  no_price  price_promo_difference  
13         0  13005399     False                  4.0096  
15         0  13835403     False                 -0.0389  
16         1  13005399     False                  4.0096  
17         0  12995397     False                  0.0036  
20         1  13005399     False                  4.0096

macbook prices seem to be mostly reasonable now. some are still way over 20000, which is to much, so I divide by 10 again.
also the ram expansions seem to be too expensive.

In [48]:
products.loc[(products.name.str.contains('MacBook')) & (products.price > 10000), 'price'] /= 10

from what i could tell, one of the highest selling synology products online seems to be around 5.500€. So I will divide every synology product with a higher price by 10.
If you look at the promo prices, you sometimes see a promoprice that is not even a 10th of the price column. This also indicates that prices are corrupted

In [49]:
products.loc[products.name.str.contains('Synology')].sort_values('price', ascending=False).head(100)

sku                                               name  \
16001  PAC2309  Synology DS1817 + | 16GB RAM | 96TB (8x12TB) S...   
16000  PAC2307  Synology DS1817 + | 8GB RAM | 96TB (8x12TB) Se...   
16004  PAC2310  Synology DS1817 + | 16GB RAM | 80TB (8x10TB) W...   
16003  PAC2308  Synology DS1817 + | 8GB RAM | 80TB (8x10TB) WD...   
15993  PAC1984  Synology DS1817 + | 16GB RAM | 80TB (8x10TB) S...   
...        ...                                                ...   
15999  PAC2305  Synology DS1817 + | 2GB RAM | 96TB (8x12TB) Se...   
1764   PAC1810  Synology DS1815 Pack + | 16GB RAM | WD 64TB Ne...   
16002  PAC2306  Synology DS1817 + | 2GB RAM | 80TB (8x10TB) WD...   
1654   PAC1799  Synology DS1815 Pack + | 8GB RAM | WD 64TB Net...   
1648   PAC1792                Synology DS1815 + | WD 64TB Network   

                                                    desc      price  \
16001       NAS with 16GB of RAM and 96TB for Mac and PC  51577.892   
16000           NAS with 8GB RAM and 96TB for Mac and PC  50917.885   
16004  NAS server with 16GB of RAM and 80TB for Mac a...  46698.692   
16003    NAS server with 8GB RAM and 80TB for Mac and PC  46038.685   
15993  NAS server with 16GB of RAM and 80TB for Mac a...  43577.892   
...                                                  ...        ...   
15999        NAS with 2GB of RAM and 96TB for Mac and PC   4978.810   
1764   Pack Synology DS1815 + with 16GB RAM + 64TB (8...   4534.490   
16002  NAS server with 2GB of RAM and 80TB for Mac an...   4490.890   
1654   Pack Synology DS1815 + with 8GB RAM + 64TB (8x...   4476.890   
1648   DS1815 + NAS capacity 64TB (8x8TB) WD Red hard...   4377.910   

       promo_price  in_stock      type  no_price  price_promo_difference  
16001   41217.4640         0  12175397     False              10360.4280  
16000   40557.4580         0  12175397     False              10360.4270  
16004   38577.4620         0  12175397     False               8121.2300  
16003   37907.4610         0  12175397     False               8131.2240  
15993   33377.4630         0  12175397     False              10200.4290  
...            ...       ...       ...       ...                     ...  
15999    3942.7463         0  12175397     False               1036.0637  
1764     3455.7455         0  12175397     False               1078.7445  
16002    3678.7461         0  12175397     False                812.1439  
1654     3402.7463         0  12175397     False               1074.1437  
1648     3306.7461         0  12175397     False               1071.1639  

[100 rows x 9 columns]

In [50]:
products.loc[(products.name.str.contains('Synology')) & (products.price > 5500), 'price'] /= 10

we see that there are still a lot inconsistencies with the prices and promoprices, but we will keep it for now and come back later. 
Next we'll look at expansions. fyi: I'll ad the .head() so I don't have to scroll so far.

In [51]:
products.loc[(products.name.str.contains('Synology'))].sort_values('price',ascending=False).head()

sku                                               name  \
15168  SYN0165  Synology RackStation RS18017xs + NAS 12 bays S...   
15166  SYN0164         Synology RackStation RS4017xs + Nas Server   
12863  PAC1481  Synology DS216 + II | WD 4TB Network for Mac a...   
16001  PAC2309  Synology DS1817 + | 16GB RAM | 96TB (8x12TB) S...   
16000  PAC2307  Synology DS1817 + | 8GB RAM | 96TB (8x12TB) Se...   

                                                    desc      price  \
15168  Nas Format rack 12 bays with Intel Xeon 6-core...  5499.0000   
15166  Nas Format rack 16 bays with Intel Xeon proces...  5389.0000   
12863  Synology DS216 + II + 4TB (2x2TB) WD Red hard ...  5309.8960   
16001       NAS with 16GB of RAM and 96TB for Mac and PC  5157.7892   
16000           NAS with 8GB RAM and 96TB for Mac and PC  5091.7885   

       promo_price  in_stock      type  no_price  price_promo_difference  
15168    5498.9999         0  12175397     False                  0.0001  
15166    5389.0000         0  12175397     False                  0.0000  
12863    5081.7940         0  12175397     False                228.1020  
16001   41217.4640         0  12175397     False              10360.4280  
16000   40557.4580         0  12175397     False              10360.4270

these expansion kits' price seems to be in the 100s, not in the 1000s based on a google search. In my case, the ones that are too pricey have a price_promo_difference of at least 200€, so I divide those by 10.

In [52]:
products.loc[products.name.str.contains('expansion')].sort_values('price_promo_difference', ascending=False).head()

sku                                               name  \
14929  PAC1907  Crucial MX300 expansion kit 2TB SSD + 16GB 160...   
11501  PAC1251  Samsung SSD 850 expansion kit PRO 1TB + 16GB 1...   
2010   PAC0746  Kit PRO 512GB Samsung SSD expansion + 16GB 160...   
1955   PAC0784  Samsung SSD 850 expansion kit EVO 500GB + 16GB...   
14923  PAC1902  SSD Crucial MX300 expansion kit 1TB + 8GB 1600...   

                                                    desc     price  \
14929  expansion kit 2TB SSD + 16GB 1600Mhz RAM for M...  7995.192   
11501  SSD upgrade kit 1TB + 16GB 1600Mhz RAM for Mac...  6989.592   
2010   SSD upgrade kit 512GB + 16GB 1600Mhz RAM for M...  4689.592   
1955   Expansion kit SSD 500GB + 16GB 1600Mhz RAM + D...  4519.592   
14923  SSD expansion 1TB + 8GB 1600Mhz RAM for Mac mi...  4463.085   

       promo_price  in_stock  type  no_price  price_promo_difference  
14929      682.584         0  1433     False                7312.608  
11501      626.584         0  1433     False                6363.008  
2010       418.585         1  1433     False                4271.007  
1955       353.585         1  1433     False                4166.007  
14923      384.584         1  1433     False                4078.501

In [53]:
products.loc[products.name.str.contains('expansion') & (products.price_promo_difference > 200), 'price'] /= 10

Next we'll look at SanDisk and Wi-Fi, also sku REP seems suspicious (sorting by price) --> SanDisk with a price over 1000€ -> divide by 10
REP: anything above 2000 -> divide by 100, anything above 300 -> divide by 10
Wi-Fi: all above 1500 -> divide by 10

In [54]:
products.loc[products.name.str.contains('SanDisk') & products.price > 1000, 'price'] /= 10
products.loc[products.sku.str.contains('REP') & products.price > 2000, 'price'] /= 100
products.loc[products.sku.str.contains('REP') & products.price > 3000, 'price'] /= 10
products.loc[products.name.str.contains('Wi-Fi') & products.price > 1500, 'price'] /= 10

In [55]:
products.loc[products.sku.str.contains('REP')].sort_values('price', ascending=False).head()

sku                                   name  \
15401  REP0350            Full Screen Repair iPad Air   
15429  REP0339             Repair Touch Screen iPad 4   
15423  REP0348            Wi-Fi antenna repair iPad 4   
15421  REP0380             iPad Mini 3 battery repair   
15397  REP0371  Cases mini repair posterior chamber 2   

                                                    desc      price  \
15401  Repair service including parts and labor for i...  19999.001   
15429  Repair service including parts and labor for i...   9990.002   
15423  Repair service including parts and labor for i...   6999.003   
15421  Repair service including parts and labor for i...   6999.003   
15397  Repair service including parts and labor for i...   6999.003   

       promo_price  in_stock      type  no_price  price_promo_difference  
15401       199.99         0  1,44E+11     False               19799.011  
15429        99.90         0  1,44E+11     False                9890.102  
15423        69.99         0  1,44E+11     False                6929.013  
15421        69.99         0  1,44E+11     False                6929.013  
15397        69.99         0  1,44E+11     False                6929.013

In [56]:
products.loc[products.sku.str.contains('SEV')].sort_values('price', ascending=False)

sku                                               name  \
525    SEV0024      Service installation RAM + HDD + SSD Mac mini   
526    SEV0025          Service installation RAM + HDD + SSD iMac   
518    SEV0026  Service installation RAM + HDD + SSD MacBook /...   
540    SEV0032                Budget Request repair Apple Mac Pro   
539    SEV0031               Budget Request repair Apple Mac mini   
538    SEV0030                   Apple iMac repair Budget Request   
537    SEV0029                Apple MacBook repair Budget Request   
536    SEV0028            Budget Request repair Apple MacBook Pro   
528    SEV0027            Budget Request repair Apple MacBook Air   
11667  SEV0041  Current 2012-Service installation RAM + HDD + ...   
517    SEV0012      HDD or SSD installation service MacBook / Pro   
524    SEV0023            HDD or SSD installation service Mac Pro   
523    SEV0022               MacBook Air SSD installation service   
522    SEV0021        SSD installation service MacBook Pro Retina   
521    SEV0020                   Installation Service Mac Pro RAM   
520    SEV0019                  Installation Service Mac mini RAM   
519    SEV0018                      IMac RAM installation service   
516    SEV0007             RAM installation service MacBook / Pro   

                                                    desc    price  \
525    installation RAM HDD + SSD + on your Mac mini ...  599.918   
526    installation RAM HDD + SSD + on your iMac + Da...  599.918   
518    RAM + HDD installation + SSD in your MacBook /...  599.918   
540                         Diagnosis to repair Mac Pro.  199.892   
539                        diagnosis to repair Mac mini.  199.892   
538                            diagnosis to repair iMac.  199.892   
537                            MacBook repair diagnosis.  199.892   
536                     diagnosis to repair MacBook Pro.  199.892   
528                     diagnosis to repair MacBook Air.  199.892   
11667  installation RAM HDD + SSD + on your iMac + Da...   99.990   
517    HDD or SSD installation on your MacBook / Pro ...   49.990   
524    HDD or SSD installation on your Mac Pro + Data...   49.990   
523    SSD installation on your MacBook Air + Data Tr...   49.990   
522    Installing SSD in your MacBook Pro Retina + Da...   49.990   
521                 Installation of RAM in your Mac Pro.   19.990   
520                      RAM installed in your Mac mini.   19.990   
519                    Installation of RAM in your iMac.   19.990   
516           Installation of RAM in your MacBook / Pro.   19.990   

       promo_price  in_stock      type  no_price  price_promo_difference  
525       599.9180         0  20642062     False                  0.0000  
526       599.9180         0  20642062     False                  0.0000  
518       599.9180         0  20642062     False                  0.0000  
540       199.8920         0  1,46E+11     False                  0.0000  
539       199.8920         0  1,46E+11     False                  0.0000  
538       199.8920         0  1,46E+11     False                  0.0000  
537       199.8920         0  1,46E+11     False                  0.0000  
536       199.8920         0  1,46E+11     False                  0.0000  
528       199.8920         0  1,46E+11     False                  0.0000  
11667      99.9896         0  20642062     False                  0.0004  
517        49.9851         0  20642062     False                  0.0049  
524        49.9851         0  20642062     False                  0.0049  
523        49.9851         0  20642062     False                  0.0049  
522        49.9851         0  20642062     False                  0.0049  
521        19.9892         0  20642062     False                  0.0008  
520        19.9892         0  20642062     False                  0.0008  
519        19.9892         0  20642062     False                  0.0008  
516        19.9892         0  2064206

In [57]:
products.loc[products.promo_price > products.price, 'promo_price'] = products.price

In [58]:
products.drop(columns = ['no_price', 'price_promo_difference'], inplace=True)

## Brands

As mentioned in the initial_exploration, the brands csv looks fine. 

## Saving the dataframes
Now we save the cleaned csv-files

In [61]:
orders.to_csv('data/orders_clean.csv', index=False)
orderlines.to_csv('data/orderlines_clean.csv', index=False)
products.to_csv('data/products_clean.csv', index=False)